### ISSUE TITLE
Save State

### ISSUE THREAD

**Issue by JackTheRipper47** [Vita3K/Vita3K#2494]
Hi, not really an issue but didn't know where else to post. Is it possible to add save states to the emulator? similar to PCSX2 or PPSSPP, that would be great if crashes happen as well cause you could just load..

anyway thanks in advance.

**Comment by CoffeeBrewer64** [comment]
the problem with save states for vita3k is that the ram is 512mb. plus some other data (CPU instructions, GPU state, etc.).

so, every single save state would be over 512mb in a single file. one of those might be okay, but the 2 or 3 states people keep on other emulators would be too much to use.

also, not everyone has an SSD. hard drives have to use a needle and some other stuff, which will always be slower than an SSD, which uses electrons and metal.

also also, android has some kind of limit on how large files can be from one call (or something like that, it might be in older versions)

**Comment by JackTheRipper47** [comment]
Oh, allright if it's that big it makes sense 😅. Though it would be cool
maybe as a pc only feature further down the line with 1 or 2 save state
slots

**Comment by CoffeeBrewer64** [comment]
agreed, but we would probably have to wait for faster and larger storage devices and ram to become more mainstream. give it a few years and save states might be a thing.

**Comment by nn9dev** [comment]
I dunno about YEARS...

There's a taiHEN plugin that implements save states on a physical Vita, which is pretty impressive. The downside though is that yes, they do take up 500+ MB per save state, and they take about ~30-60s to create and save each save state. TaiHEN plug-ins aren't supported on Vita3K, however, and they probably won't be for a bit.

Theoretically, someone could implement save states without it being TOO grueling of a process (I mean, save states are practically a glorified RAM dump), but things are constantly changing and being updated in the emulator, thus producing inconsistencies. The goal would be to have as many inconsistencies worked out as possible before implementing something that relies on. Well. Consistency. 

Not to mention there's a lot of other stuff that still needs to be worked on: graphics, audio, etc.

TL;DR, entirely possible, probably not soon.

**Comment by mcabel** [comment]
one "workaround" the 512mb problem, is a configuration option that lets you choose how many savestate slots the user want available, and maybe separate that disk space (with a dummy 512mb savestate or something), so the user will always have enough disk available for the amount of savestate slots they chose.  For most people, 1024mb , for 2 slots is probable not a deal breaker I believe.

**Comment by CoffeeBrewer64** [comment]
True, but what about read write speeds? not everyone has an SSD. and the file size of 512mb would still be present. but a dummy file is a good idea

**Comment by nn9dev** [comment]
There's honestly no need for any of that. A check before creating the save state for storage makes that a non-issue. And it's far less of an issue, and more of simply an "annoyance", that each save state has the potential to be the size of a whole game. 

Additionally, a SSD isn't even really needed. Most hard drives these days have write speeds of upwards of 150MB/s, (so, ~3.5s to write 512MB) and even if you have an older one you're probably still not really gonna suffer much from ONE 512MB write. Even your average 2011 hard drive begins to boast ~70MB/s+ (~7.3s to write 512MB)

So again, the limitations truly aren't physical or even technical, but more that it's just a bit of an annoyance (and that it hasn't been done yet (and that there's just some other focuses)

**Comment by Macdu** [comment]
> the problem with save states for vita3k is that the ram is 512mb. plus some other data (CPU instructions, GPU state, etc.).
> 
> so, every single save state would be over 512mb in a single file. one of those might be okay, but the 2 or 3 states people keep on other emulators would be too much to use.
> 
> also, not everyone has an SSD. hard drives have to use a needle and some other stuff, which will always be slower than an SSD, which uses electrons and metal.
> 
> also also, android has some kind of limit on how large files can be from one call (or something like that, it might be in older versions)


That's kinda wrong, the PS Vita indeed has 512 MB of RAM. However, most games do not use half of it. Moreother, a good chunk of the memory used by the PS Vita is not used by Vita3K because of HLE emulation (no need for many low level files  and many graphics buffers). Finally, compressing the data before writing it would be really effective as most of what takes space has a really low entropy. The PS3 has 512MB of memory and RPCS3 has save states.

The main issue is that it's really hard and annoying to implement and make it right (you basically have to serialize/unserialize the whole emulator state)

**Comment by memory-hunter** [comment]
Where RPCS3 managed to implement it, it think it is viable for Vita3K too

> the problem with save states for vita3k is that the ram is 512mb. plus some other data (CPU instructions, GPU state, etc.).
> 
> so, every single save state would be over 512mb in a single file. one of those might be okay, but the 2 or 3 states people keep on other emulators would be too much to use.
> 
> also, not everyone has an SSD. hard drives have to use a needle and some other stuff, which will always be slower than an SSD, which uses electrons and metal.
> 
> also also, android has some kind of limit on how large files can be from one call (or something like that, it might be in older versions)

**Comment by jonshipman** [comment]
I think a single save state would be fine. Like being able to take a break from a dungeon in P4G without using a seed to return to the entrance to save.

Less treating it like a way to cheese the game and more like taking a break. Akin to locking a Vita and stowing it away in your bag.

**Comment by Phil-T1** [comment]
This would be amazing, the emulator works really well, but for me a big thing about playing the good old games is being able to stop and start as required.

Others have mentioned crashing, but even without, playing on my phone and having a busy life means I'll likely abandon more frequently than reach a savepoint 😀

Choosing an amount of storage to dedicate seems perfectly fine 🙂

**Comment by paulo101977** [comment]
> I dunno about YEARS...
> 
> There's a taiHEN plugin that implements save states on a physical Vita, which is pretty impressive. The downside though is that yes, they do take up 500+ MB per save state, and they take about ~30-60s to create and save each save state. TaiHEN plug-ins aren't supported on Vita3K, however, and they probably won't be for a bit.
> 
> Theoretically, someone could implement save states without it being TOO grueling of a process (I mean, save states are practically a glorified RAM dump), but things are constantly changing and being updated in the emulator, thus producing inconsistencies. The goal would be to have as many inconsistencies worked out as possible before implementing something that relies on. Well. Consistency.
> 
> Not to mention there's a lot of other stuff that still needs to be worked on: graphics, audio, etc.
> 
> TL;DR, entirely possible, probably not soon.

For me, save state is a crucial feature since I use emulators to train Reinforcement Learning agents. If this feature ever becomes available, I'll get my hands on the Vita3k code and port it to meet my needs.

**Comment by GSTDiandre** [comment]
Savestates are vital for the emulator to be viable in the long run for a plethora of reasons so hope this is being considered.
